test/compress/pure_funcs.js

```diff
@@ -0,0 +1,55 @@
+array: {
+    options = {
+        pure_funcs: [ "Math.floor" ],
+        side_effects: true,
+    }
+    input: {
+        function f(a, b) {
+            Math.floor(a / b);
+            Math.floor(c / b);
+        }
+    }
+    expect: {
+        function f(a, b) {
+            Math.floor(c / b);
+        }
+    }
+}
+
+func: {
+    options = {
+        pure_funcs: function(node) {
+            return !~node.args[0].print_to_string().indexOf("a");
+        },
+        side_effects: true,
+    }
+    input: {
+        function f(a, b) {
+            Math.floor(a / b);
+            Math.floor(c / b);
+        }
+    }
+    expect: {
+        function f(a, b) {
+            Math.floor(c / b);
+        }
+    }
+}
+
+side_effects: {
+    options = {
+        pure_funcs: [ "console.log" ],
+        side_effects: true,
+    }
+    input: {
+        function f(a, b) {
+            console.log(a());
+            console.log(b);
+        }
+    }
+    expect: {
+        function f(a, b) {
+            console.log(a());
```

Comment:
That isn't an issue per se, because it makes sense when asking `AST_Node.has_side_effects()` to consider the expression as a whole and return `true` if any of its components have side effects.

Optimisation of individual side-effect-free components are scattered across `compress.js`, and in this particular case, I think `OPT(AST_SimpleStatement)` needs to be augmented as it's the site where we know the evaluated value of its `body` is being discarded.